Add Differential Calculus methods

aktech opened the pull request:
TODO
- [x] is_increasing
- [x] is_strictly_increasing
- [x] is_decreasing
- [x] is_strictly_decreasing
- [x] is_monotonic
- [x] Doc strings
- [x] Reference in Sphinx docs (already referenced)
- [x] Tests
- [x] Error message with `NotImplementedError`
- [x] Merge master after #9822 gets Merged
- [x] Update release Notes.

@hargup @flacjacket

aktech committed:
Add Differential Calculus methods

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

aktech commented:
@moorepants 
What do you think about the function names, Is that good?

moorepants commented:
Seems fine, can `given_interval` just be `interval`?

moorepants commented on sympy/calculus/singularities.py:
```diff
@@ -32,3 +33,123 @@ def singularities(expr, sym):
+def is_increasing(f, given_interval):
+    """
+    Returns if a function is increasing or not, in the given
+    `Interval`.
```
Does it actually tell you both? If it only tells you the function is increasing, then the function name is fine, but if this tells you whether it is increasing or decreasing, then maybe you want a function like: `increasing_or_decreasing`. I don't think it is necessarily true that a function is decreasing if it is not increasing. It could be neither.

aktech replied:
It tells if a function is:

* increasing
* not increasing (can be anything: decreasing, (both increasing & decreasing) or neither)

asmeurer replied:
Maybe remove the "or not". It's redundant and confusing.

aktech committed:
Add tests for differential calculus methods

aktech committed:
Replace `given_interval` -> `interval

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

aktech committed:
`df_pos_interval` -> `df_nonneg_interval`

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

debugger22 commented:
Showing a message would be good in not implemented exceptions.

hargup commented:
@aktech please also add the reference of these functions in the main sphinx documentation.

aktech committed:
Add Error message for NotImplementedError

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

aktech commented:
> please also add the reference of these functions in the main sphinx documentation.

Currently automodule is set in the sphinx documentation for `singularities.py`, so all the functions in the module is automatically added, [See here](https://raw.githubusercontent.com/sympy/sympy/master/doc/src/modules/calculus/index.rst).

aktech commented:
> Showing a message would be good in not implemented exceptions.

Yes, thanks.

aktech committed:
Add is_monotonic function

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

aktech committed:
Replace `assert not foo` with `assert foo is False`

This is done basically due to the fact `not foo` returns
`True` in either case when `foo` is `False` or `None`

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

aktech commented:
Update: `is_monotonic` Added.

aktech committed:
Fix Typo in is_monotonic function

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

gxyd commented on sympy/calculus/singularities.py:
```diff
@@ -1,6 +1,7 @@
 from sympy.solvers import solve
```
This import can be removed ?

aktech replied:
Thanks!

gxyd commented on sympy/calculus/singularities.py:
```diff
@@ -32,3 +33,160 @@ def singularities(expr, sym):
+
+    >>> from sympy.calculus.singularities import is_strictly_increasing
+    >>> from sympy.abc import x
+    >>> from sympy import S, Interval, oo
```
`S` has not been used in docstring

aktech replied:
Thanks!

gxyd commented on sympy/calculus/singularities.py:
```diff
@@ -32,3 +33,160 @@ def singularities(expr, sym):
+    Examples
+    ========
+
+    >>> from sympy.calculus.singularities import is_increasing
```
This import can be replaced with `from sympy import is_increasing`

aktech replied:
Thanks!

gxyd commented on sympy/calculus/singularities.py:
```diff
@@ -32,3 +33,160 @@ def singularities(expr, sym):
+
+    >>> from sympy.calculus.singularities import is_monotonic
+    >>> from sympy.abc import x
+    >>> from sympy import S, Interval, oo, Or
```
`Or` not used

aktech replied:
Thanks!

asmeurer commented on sympy/calculus/singularities.py:
```diff
@@ -32,3 +33,160 @@ def singularities(expr, sym):
+def is_increasing(f, interval):
+    """
+    Returns if a function is increasing or not, in the given
+    `Interval`.
```
One set of backticks is math. You need to use two for code.

asmeurer commented:
Should interval default to S.Reals?

aktech committed:
Remove unused imports and fix doc formatting

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

aktech commented:
> Should interval default to S.Reals?

I think that would be good.

aktech committed:
Add default value of interval to S.Reals

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

aktech committed:
remove unused import

aktech committed:
Minor doc formatting

aktech committed:
UnXFAIL some tests and doctests

Signed-off-by: AMiT Kumar <dtu.amit@gmail.com>

hargup commented:
+1
